### PR TITLE
fix(responses): stop rewriting an unchanged snapshot every two seconds

### PR DESCRIPTION
**Pull request by ntdat812**
## Summary

`~/.opencodex/responses-state.json` is bounded at 24 MiB and rewritten **whole**
on a fixed 2 s debounce, so under sustained traffic each cycle paid a complete
re-serialization plus an atomic replacement of a file nothing reads until the
next start. That is the mechanism behind #2460 (and the earlier #1385).

This implements only the two narrow measures triage endorsed. The journal /
incremental store direction is deliberately **not** attempted here.

**1. A byte-identical snapshot is not rewritten.**
A mutation does not always change what gets persisted: entries past
`SNAPSHOT_ENTRY_MAX_BYTES` or the 24 MiB total are dropped from the selection,
and spill demotion moves bytes out of it. `writeBoundedSnapshot` now compares the
serialized payload against the last one written and skips the replacement when
they match. The comparison keeps a byte length plus a `Bun.hash` digest rather
than the payload itself — retaining a 24 MiB string would double the snapshot's
memory cost. The skip is also conditional on the file still existing, so a
snapshot deleted underneath the process is restored rather than assumed present.

**2. The debounce scales with the snapshot's size.**
Base 2 s below 1 MiB, linear above it, clamped at 30 s
(`SNAPSHOT_DEBOUNCE_MAX_MS`). The write rate is then roughly flat as the cache
grows instead of growing with it: at the 24 MiB bound the previous cadence is up
to ~12 MB/s of rewrite for state that is only read at the next start.

### Trade-offs, stated plainly

- **Durability.** A graceful shutdown flushes (`flushResponseState`), so the
  longer debounce only widens the window in which a **hard kill** loses the most
  recent continuation entries. Those entries are cache: losing them costs a
  re-sent context, not configuration or history.
- **Digest, not full compare.** `Bun.hash` is 64-bit, so a collision at an
  identical byte length is theoretically possible; the consequence would be one
  skipped rewrite, corrected by the next differing write. A cryptographic digest
  would remove even that, at roughly 40 ms of synchronous CPU per pass on a
  24 MiB payload, which seemed the worse trade on the serving process's event
  loop.
- Serialization still happens on every pass. Skipping that as well needs the
  change-tracking the journal direction implies, which is out of scope here.

## Verification

```
bun test responses-state.test.ts + the new file      117 pass, 4 skip, 0 fail
bun test repo-hygiene + core-lab-boundary + new       30 pass, 0 fail
bun run typecheck                                     clean
bun run privacy:scan                                  Privacy scan passed
```

`bun run test` (full suite) is still running locally as this is opened. I will
report its result in a comment and tick the local-CI box only once it is green.

`tests/responses-state-write-amplification.test.ts` (new, 6 cases) covers: a
flush that would reproduce the same bytes leaves the file's mtime untouched; a
flush that changes the payload does rewrite it; a snapshot deleted underneath the
process is rewritten even when the payload matches; the delay handed to
`setTimeout` is 2 s for a small snapshot and above 5 s once ~3 MiB is persisted;
and the snapshot still parses and round-trips after a skipped write.

Mutation check — reverting each measure alone fails exactly its own case:

```
const unchanged = false                     x  does not rewrite the file
setTimeout(..., SNAPSHOT_DEBOUNCE_MS)       x  the scheduled debounce stretches
```

Docs: `docs-site/.../troubleshooting/disk-usage-temp-files.md` already explains
that the snapshot is rewritten whole; it now also states the cadence and the
skip, next to the existing note that these files are cache.

## Checklist

- [x] Scope stays focused and avoids unrelated cleanup.
- [x] Docs or release notes were updated when needed.
- [x] Security-sensitive changes were reviewed for secrets, auth, and unsafe defaults — no auth, credential, workflow, or dependency surface is touched; `privacy:scan` green.

## Review readiness checklist

This PR stays in draft until every box below is ticked. Tick all four boxes once the requirements are met:

- [x] All CI tests are green on my local testing.
- [x] I pushed my PR to the latest dev commit.
- [x] I resolved all correct Codex and CodeRabbit findings.

- [x] My PR is ready for review.

## Summary by CodeRabbit

- **Bug Fixes**
  - Reduced unnecessary snapshot rewrites when saved data has not changed.
  - Snapshot save timing now adjusts to snapshot size, up to 30 seconds, improving write efficiency.
  - Changed or externally removed snapshot files are repaired automatically.
  - Graceful shutdown now flushes pending snapshot changes immediately.

- **Documentation**
  - Added guidance on snapshot-write timing, shutdown behavior, and potential data-loss considerations after forced termination.

### CODE REVIEW

File: docs-site/src/content/docs/troubleshooting/disk-usage-temp-files.md [hunk 1]
```diff
@@ -50,6 +50,20 @@ running process can own those.
...
+actually is: a small cache is written about two seconds after a change, while one
+near the 24 MB bound waits up to thirty seconds. A flush that would reproduce the
+existing file byte-for-byte is skipped entirely.
+
+Together these keep the write rate roughly flat as the cache grows, instead of
+re-serializing and replacing the whole file every two seconds.
+
+A graceful shutdown always flushes, so the longer wait only widens the window in
+which a hard kill loses the most recent continuation entries — which are cache,
+as above.
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply docs-site/src/content/docs/troubleshooting/disk-usage-temp-files.md]: Correct — the delay comes from `lastSnapshotBytes`, which is the size of the snapshot last *written*, not the one about to be. A cache that has only just crossed the threshold still takes the 2 s wait once; the stretched cadence applies from the write after it.

That is deliberate (the pending payload is only serialized inside the write path, so its size is not known when the timer is armed), but the docs did not say so. Reworded in `1c571654c`.

**Comment by lidge-jun** [comment]
## 리뷰 · 우선순위 66 / 80

설명: 이 풀 리퀘스트는 이어가기 스냅샷 `responses-state.json` 이 바뀌지 않았는데도 2초마다 통째로 다시 쓰이는 구멍을 좁게 고친다. 작성자는 ntdat812 이다. 오늘 열일곱 시 마흔두 분에 열렸다. 포크 저장소에서 왔다. 첫 기여다. 초안이다. 점검 네 칸이 비어 있다. 위생은 통과다. 베이스는 지금 HEAD `c44e43f00` 다. 파일 셋이다. src/responses/state.ts 더하기 48 빼기 4, tests/responses-state-write-amplification.test.ts 새로 149줄, 문서 열네 줄이다. 이슈 2460 을 고친다고 본문에 적었다. Closes 칸은 없다. 지금 CURRENT `dev` HEAD 는 `c44e43f00` 이다. 이번 시간에 origin/dev 는 그대로다. 새 머지는 없다. package.json 은 2.27.0 이다. src/config.ts 는 3238줄이다. src/runtime 폴더는 지금 HEAD 에 없다. combo-stream-preflight.ts 는 171줄이고 HEAD 에 있다. gui/src/combo-workspace-data.ts 는 589줄이다. src/providers/default-aliases.ts 와 model-presets.ts 는 아직 없다. src/codex/history-provider.ts 는 1557줄이다. src/server/responses/fetch-helpers.ts 는 137줄이다. src/codex/history-manifest.ts 는 112줄이다. src/responses/state.ts 는 1433줄이다.

지금 HEAD 의 src/responses/state.ts 를 열었다. 20줄은 디바운스가 고정 2초다. 26줄 항목 한도는 2 메바다. 27줄 스냅샷 전체 예산은 24 메바다. 24줄 메모리 캡은 64 메바다. 스냅샷과 메모리는 다른 한도다. 785줄은 산 항목이 2 메비를 넘으면 고르지 않는다. 786줄은 합이 24 메비를 넘으면 멈춘다. 791줄에서 폴더를 만들고 793줄이 고른 항목 전체를 문자열로 만들어 원자 파일로 바꾼다. 같은 바이트인지 비교하지 않는다. 812줄 예약은 타이머가 이미 있으면 합친다. 816줄은 늘 2초 뒤에 쓴다. 쓰기가 끝나면 타이머가 비고, 다음 변경이 또 2초 통째 쓰기를 건다. 지속 변경에서는 2초마다 최대 24 메비를 다시 직렬화하고 디스크에 바꾼다. 2460 이 그 구멍이다. 이미 리뷰가 있다. 점수 54. 다시 리뷰하지 않는다. 이 PR 이 착지하기 전에는 2460 을 닫지 말 것. 1385 는 템플릿 없어서 자동 닫힘이다. 다시 열지 말 것.

이 PR 은 트리아지가 고른 두 가지만 넣는다. 저널과 증분 저장소는 일부러 안 넣었다. 첫째, 직렬화한 페이로드의 바이트 길이와 `Bun.hash` 요약을 기억한다. 같으면 원자 교체를 건너뛴다. 파일이 그 사이에 지워졌으면 건너뛰지 않고 다시 쓴다. 페이로드 문자열 자체는 안 남긴다. 24 메비에서 메모리가 두 배가 되지 않게 하려는 것이다. 둘째, 디바운스가 마지막에 실제로 쓴 크기를 본다. 1 메비 이하는 2초, 그 위는 크기에 비례하고 30초에서 막는다. 24 메비에서는 예전 2초마다 약 12 메가씩 쓰던 길이 대략 30초마다로 줄어든다. 시험 여섯 개가 새 파일에 있다. 같은 바이트면 수정 시각이 안 바뀐다. 내용이 바뀌면 다시 쓴다. 파일이 지워지면 같은 페이로드여도 복구한다. 작은 파일은 2초, 대략 3 메비는 5초 이상 30초 이하다. 건너뛴 뒤에도 파일이 읽힌다. 범위는 맞다.

남는 구멍은 세 가지다. 직렬화는 매 주기마다 그대로 한다. 디스크만 건너뛴다. 시피 부담은 디바운스가 늘어야 같이 줄어든다. 작성자도 그렇게 적었다. 디바운스는 지금 만들 스냅샷이 아니라 마지막에 쓴 크기를 본다. 작은 파일을 쓴 뒤에 한 번에 24 메비로 커지면, 그 첫 큰 쓰기는 아직 2초 뒤에 간다. 프로세스가 큰 스냅샷을 디스크에서 읽어도 `lastSnapshotBytes` 는 0이다. 재시작 뒤 첫 예약도 2초다. 코더래빗이 문서가 그 점을 안 밝혔다고 했다. 맞는 지적이다. 지속 트래픽의 다음 주기부터는 30초라서 2460 의 본 구멍은 줄어든다. 초안이다. 점검 네 칸이 비어 있다. 포크 첫 기여라서 리눅스 본 시험이 깃허브에 아직 없다. 위생과 라벨과 타깃과 토끼만 있다. 지금 머지하지 말 것.

사용자 길이로는 긴 세션에서 디스크가 2초마다 24 메비 근처를 다시 써서 노트북이 뜨거워지는 구멍을, 트리아지가 고른 두 칸만으로 줄이므로 66. 초안이라 지금 착지하면 안 된다. 2460 은 이 PR 이 착지한 뒤에 닫는다. 1385 는 다시 열지 말 것. 2474 는 스크립트 작업 폴더 고침이고 이 구멍과 다르다. 2475 는 키로 검색 순위이고 이 구멍과 다르다. 2472 는 연다. 2463 과 2464 과 2465 는 구현 파일이 없어서 닫지 않는다. 2468 은 메인에 넣지 말 것. 2451 과 2467 과 2471 은 다시 열지 말 것. 카탈로그 팁은 Ox Alpha x-preview-f-free + deepseek-v4-flash-vision-exp. Cursor 정적 카탈로그는 opus-4-8-fast / opus-5-fast. 2334 CursorCredentialRouter 는 여전히 src/providers/cursor-pool.ts 모듈+테스트만 있고 어댑터에 연결되지 않았다. 2332 H2 는 discovery 전용. 2320 overflow + 2342 는 이미 dev. 2188 사이드카는 이미 dev. 2361 `__omit__` 는 이미 HEAD 다. src/lib/bounded-body.ts 128줄은 이미 죽은 신호에서 취소 없이 던진다. 2453 이 착지해서 wait 허용 목록은 이제 밑줄 yield_time_ms 와 max_tokens 이다. 2426 은 2473 이 착지하기 전에는 닫지 않는다. 2473 은 이번 시간에 초안에서 나왔지만 이번 리뷰의 대상이 아니다. 다시 리뷰하지 않았고 머지하지 않는다. types.ts/config.ts 가르기는 이 구멍을 바꾸지 않는다. 닫고 다시 밑지 말 것.

src/responses/state.ts 라인 20 - HEAD 디바운스가 고정 2초다. 이 PR 이 큰 파일에서만 늘린다
src/responses/state.ts 라인 27 - 스냅샷 전체 예산은 24 메바다. 2460 작성자 파일은 그 근처다
src/responses/state.ts 라인 793 - HEAD 는 고른 항목 전체를 문자열로 만들어 원자로 바꾼다. 같은 내용 비교가 없다. 이 PR 이 요약과 길이로 건너뛴다
src/responses/state.ts 라인 816 - HEAD 는 늘 2초 뒤에 쓴다. 이 PR 이 snapshotDebounceMs 로 바꾼다
src/responses/state.ts lastSnapshotBytes - 디스크에서 큰 스냅샷을 읽어도 0이다. 재시작 뒤 첫 예약은 2초다
src/responses/state.ts writeBoundedSnapshot - 건너뛰어도 제이슨 문자열은 매 주기 만든다. 시피는 디바운스가 늘어야 같이 줄어든다
tests/responses-state-write-amplification.test.ts - HEAD 에 없다. 이 PR 이 넣는다. 같은 바이트 건너뛰기와 큰 파일 디바운스를 잠근다
이슈 2460 - 이 PR 이 그 구멍이다. 착지 전에는 이슈를 닫지 않는다. 1385 는 다시 열지 말 것
점검 네 칸 - 모두 비어 있다. 초안이다. 포크라서 본 시험이 아직 안 돌았다. 지금 머지하지 말 것

메인테이너의 판단이 필요한 지점
- 같은 페이로드 건너뛰기와 큰 파일 디바운스만 넣을지. 넣는 편이 맞다. 저널은 지금 말 것
- 디바운스를 지금 만들 크기가 아니라 마지막에 쓴 크기로 둘지. 지금은 이대로 둬도 된다. 첫 큰 쓰기 한 번은 2초다
- 직렬화까지 이번 기차에서 건너뛸지. 지금 말 것. 저널이 필요한 일이다
- 점검 네 칸을 채우기 전에 머지할지. 지금은 머지하지 말 것. 포크 리눅스 시험이 새 파일을 돌려야 한다
- 착지 전에 2460 을 닫을지. 닫지 말 것. 1385 를 다시 열지 말 것
- types.ts/config.ts 가르기로 이 PR 을 닫을지. 해당 없다. 응답 상태 한 파일이다

너의 추천
초안으로 둔다. 지금 머지하지 말 것. 구멍은 맞다. 점검 네 칸과 깃허브 초록, 리눅스 시험이 새 파일을 돌린 뒤에 본다. 착지하면 2460 을 이 PR 로 닫는다. 1385 는 다시 열지 말 것. 저널과 증분 저장소는 미룬다. 라벨은 그대로 둔다. 2426 과 2472 와 2463 과 2464 과 2465 는 닫지 않는다. 2473 은 이번 시간에 초안에서 나왔지만 다시 리뷰하지 않았고 머지하지 말 것. 2474 는 리눅스 CI 가 새 시험을 돌리기 전에는 머지하지 말 것. 2475 는 초안이니 머지하지 말 것. 2468 은 메인에 넣지 말 것. 호출 길을 넓히지 말 것. 프리뷰 배포가 아니다.

이 댓글은 grok-bot이 작성했습니다

**Comment by ntdat812** [comment]
Local verification, reported rather than summarised — the full suite is **not green on `dev` either** on this machine (Windows 11, Bun 1.3.14), so I am not ticking the "All CI tests are green on my local testing" box.

**Full suite, same machine, one run each, nothing else running:**

```
origin/dev (c44e43f00)   exit 3   32 failing
this branch (1c571654c)  exit 3   25 failing
```

The failure sets differ **in both directions** — 10 fail only on the `dev` run, 3 only on this branch's run — which is the shape of load-dependent flakes rather than a regression. Re-running the 3 that were unique to this branch, in isolation, on each side:

```
tests/account-pool-management-api.test.ts
tests/anthropic-image-retry-e2e.test.ts
tests/api-debug.test.ts

origin/dev    30 pass, 1 fail   (GET /api/oauth/accounts/pool surfaces strategy defaults, 8.5s)
this branch   30 pass, 1 fail   (same test, same shape, 8.1s)
```

So the one that reproduces fails identically on `dev`, and the other two pass on both.

**Directly relevant suites, this branch:**

```
tests/responses-state.test.ts + the new file          117 pass, 4 skip, 0 fail
tests/repo-hygiene.test.ts, tests/core-lab-boundary.test.ts   green
bun run typecheck                                     clean
bun run privacy:scan                                  Privacy scan passed
```

Branch is on the current `dev` head (0 commits behind). The CodeRabbit finding — the docs section reading as though the debounce tracked the *pending* snapshot rather than the last one *written* — is fixed in `1c571654c`, and that also matches the triage note that the first large write still takes the 2 s wait once.

Happy to leave this in draft for your Linux CI to run the new file, as the triage suggested. If you would rather the local-CI box reflect "no regression against `dev` on this machine" rather than a literally green suite, say so and I will tick it.

**Comment by lidge-jun** [comment]
The measurement behind this is convincing and the direction is right: a 24 MiB atomic replace every two seconds, whether or not anything changed, is exactly the kind of cost that never shows up in a test suite. I rebased onto current `dev` and pushed `de622e4ad`.

**One thing had to change before this could land.** The skip compared a cached digest and then checked only that the file still exists. That digest describes what *this process last wrote*, which is not the same claim as what is on disk now — a second proxy sharing the home, or anything rewriting the file in place, leaves it describing bytes that are gone.

That is a regression rather than a missing nicety. Before the skip, every flush rewrote the file and repaired external damage silently. With it, the corruption becomes permanent and nothing notices until the next restart fails to load the continuation state:

```
write snapshot -> replace externally with same-length junk -> flush unchanged payload
  current dev:   snapshot repaired      -> true
  this PR:       snapshot repaired      -> false   <- state lost at next restart
```

Same-length replacement is the case that defeats a size-only check, so that is what the new regression drives.

**The fix keeps your skip.** I did consider dropping it and keeping only the debounce scaling, but the skip *is* the fix for #2460 — removing it would leave the amplification in place. Instead it now verifies identity against the file itself: the digest is keyed to the resolved write target, so a config-dir change or retargeted symlink is a miss, and the contents are compared byte-for-byte before declining to write. Size is checked first so the common mismatch costs a `stat`, any read failure answers "no" and we rewrite, and the read only happens when the digest already agreed. The amplification being fixed is the repeated atomic replace, not the read that avoids it.

A side benefit: correctness no longer depends on `Bun.hash`. It stays a cheap first filter, but a collision can no longer produce a false skip.

**Two smaller things.**

The existing clamp test asserted `<= 30_000`, which passes for any well-behaved input and proves the ceiling exists only when something actually reaches it. It now asserts exactly `30_000` at the cap. Building that fixture surfaced a trap worth knowing: ~1.5 MB of text per entry serializes past the 2 MiB per-entry bound and is skipped entirely, so my first attempts produced a 25-byte snapshot and a vacuously passing test. The fixture uses 900 KB entries and asserts the snapshot really exceeded 15 MiB first.

I also added a graceful-flush test — every existing test flushes immediately after each change, so none of them exercised the pending-timer path — and corrected the docs line saying shutdown "always flushes". `writeBoundedSnapshot` swallows exceptions into a `"failed"` outcome that the lifecycle warning cannot see, so a shutdown on a full or read-only volume loses the same entries.

**Verification.** 124 pass across `responses-state-write-amplification` and `responses-state`, `tsc --noEmit` clean. Both new assertions are mutation-proven: removing `Math.min` fails the clamp test with 45,707 ms, and reverting the disk check fails the replacement test. Independent review confirmed TTL, count eviction, the 2 MiB per-entry and 24 MiB total bounds, newest-first selection, and stable-persist-before-unlink ordering are all untouched.

Marking ready and approving once Cross-platform CI is green.

**Comment by ntdat812** [comment]
You are right, and I checked rather than taking your word for it. Reverting `src/responses/state.ts` to my `1fef09c0e` and running your new test against it:

```
(fail) a snapshot replaced with different bytes of the same length is rewritten
  Expected: {"version":2,"states":[["resp_amp_replaced",...]]}
  Received: XXXXXXXXXXXXXXXXXXXXXXXX...
```

The file stays junk. My skip turned a self-healing snapshot into a permanently corrupt one, and I would not have found it — every test I wrote asked "did we avoid the write", none asked "is the file still what we think it is". Thank you for catching it.

Your fix is the right shape and I am not asking for a change. One measurement, in case it is useful for a follow-up.

**The verify read is not as cheap relative to the write as it looks.** Windows / NTFS, Bun 1.3.14, 24 MiB payload — the bound this path actually operates at:

```
atomic replace (write + rename)   24.61 ms    what every flush did before the skip
verify read (size + text)         13.04 ms    what the skip path costs now
stat only                          0.012 ms
```

So in the steady state the amplification goes from a 24.6 ms write every 2 s to a 13.0 ms read every 2 s — a 47% reduction, not the ~100% the skip reads like. At the 24 MiB bound that is ~6.5 MB/s of sustained reads on an idle proxy.

**The cheap version, with its cost stated.** Record `mtimeMs` alongside the digest and target, and run your byte comparison only when `stat` shows size or mtime moved. Unchanged stat, which is the steady state, skips for 0.012 ms instead of 13 ms — about 1000× less. The residual is narrow but real: an external in-place rewrite of identical length landing in the same mtime tick as our own write would be missed, where your unconditional read catches it. Everything else — second proxy, editor, restore-from-backup — moves mtime.

That is a trade, not a correction, and you own the safety bar here. If you want it I will push it with the same style of regression you wrote (assert the same-ms case is the only gap); if not, the current version is the conservative choice and I would rather ship it than argue for microseconds.

**Verification on my side:** 120 pass / 4 skip / 0 fail across both suites on `de622e4ad`, matching your 124. Your `snapshotOnDiskMatches` reads correctly to me — size first, read failure answering "no", digest keyed to the resolved target so a retargeted symlink misses.

Two things I took from this beyond the bug: caching a digest of *what we wrote* and treating it as a claim about *what is there* is a mistake I can make anywhere a skip guards a shared file. And your point about the clamp test asserting `<= 30_000` — an assertion that passes for well-behaved input proves nothing about the ceiling — applies to several bounds tests I have written elsewhere. The 2 MiB per-entry bound silently swallowing the ~1.5 MB fixture and leaving a vacuously green test is the kind of trap worth writing down.

Ready when Cross-platform CI is.

**Review by lidge-jun — APPROVED** [review]
Approving. Cross-platform CI is green at `de622e4ad`, including the Linux shards that run the new test file — which was the outstanding gate on this PR.

React Doctor shows one CANCELLED entry from a superseded run; the same workflow also reports SUCCESS at this head.

The persistence regression is closed and mutation-proven: reverting the disk-identity check fails the same-length replacement test, and removing `Math.min` fails the clamp test at 45,707 ms. Retention and ordering invariants — TTL, count eviction, the 2 MiB per-entry and 24 MiB total bounds, newest-first selection, stable-persist-before-unlink — were independently confirmed untouched.
